### PR TITLE
Add `jsifyMapListProp` and `unjsifyMapListProp` conversion utils

### PR DESCRIPTION
**Pull request by sydneyjodon-wk**
## Motivation
  

We need prop conversion utils to convert and unconvert `List<JsMap>` using `(un)jsifyMapProp` utils for [DataGridPro generation](https://github.com/Workiva/react_material_ui/pull/376#discussion_r1039911301)

## Changes
  

- Add `jsifyMapListProp` and `unjsifyMapListProp` conversion utils
- Add tests 

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
      - [ ] Verify good test coverage
      - [ ] CI passes
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/util/prop_conversion.dart [hunk 1]
```diff
@@ -51,6 +51,16 @@ Map<K, V> unjsifyMapProp<K, V>(JsMap value) {
...
 }
 
+/// Runs [jsifyMapProp] on every Dart Map in the [value].
+List<JsMap> jsifyMapListProp(List<Map> value) {
+  return value?.map(jsifyMapProp)?.toList();
+}
+
+/// Runs [unjsifyMapProp] on every JS Map in the [value].
+List<Map<K, V>> unjsifyMapListProp<K, V>(List<JsMap> value) {
+  return value?.map(unjsifyMapProp)?.toList()?.cast();
```
Comment (greglittlefield-wf) [inline lib/src/util/prop_conversion.dart:61]: I think the `cast` here might not be necessary
```suggestion
  return value?.map(unjsifyMapProp)?.toList();
```

File: test/over_react/util/prop_conversion_test.dart [hunk 2]
```diff
@@ -75,16 +73,103 @@ main() {
...
+              ]),
+              equals([
+                {'foo': 'bar'},
+                {'bar': 1},
+                {'foo2': true},
+              ]),
+            );
+          });
+
+          test('casting the returned map to the correct type', () {
```
Comment (greglittlefield-wf) [inline test/over_react/util/prop_conversion_test.dart:152]: Since we're verifying the list itself here
```suggestion
          test('casting the returned list to the correct type', () {
```

File: test/over_react/util/prop_conversion_test.dart [hunk 3]
```diff
@@ -75,16 +73,103 @@ main() {
...
+              {'bar': 1},
+              {'foo2': true},
+            ]),
+            orderedEquals([
+              hasJsBackedMapValue({'foo': 'bar'}),
+              hasJsBackedMapValue({'bar': 1}),
+              hasJsBackedMapValue({'foo2': true}),
+            ]),
+          );
+        });
```
Comment (greglittlefield-wf) [inline test/over_react/util/prop_conversion_test.dart:107]: Could we add a test around how `null` entries in the list are handled?

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Couple small things, plus the additional test coverage from that other thread, otherwise this looks good!

File: lib/src/util/prop_conversion.dart (hunk 1 above)
Comment (sydneyjodon-wk) [reply lib/src/util/prop_conversion.dart:61]: I'm getting this error without the `cast`:
```
A value of type 'List<Map<dynamic, dynamic>>' can't be returned from the function 'unjsifyMapListProp' because it has a return type of 'List<Map<K, V>>'.
```

Comment (sydneyjodon-wk) [reply lib/src/util/prop_conversion.dart:61]: Oh wait I guess it works if I take out that `<K, V>` - is that what you were thinking?

File: test/over_react/util/prop_conversion_test.dart [hunk 4]
```diff
@@ -75,16 +73,111 @@ main() {
...
+              ]),
+            );
+          });
+
+          test('casting the returned list to the correct type', () {
+            expect(
+                unjsifyMapListProp<String, String>([
+                  jsify({'foo': 'bar'}) as JsMap
+                ]),
+                isA<List<Map<String, String>>>());
```
Comment (greglittlefield-wf) [inline test/over_react/util/prop_conversion_test.dart:165]: Related to the comment about `.cast`, could we also add checks here that the elements are the correct type, and not just the container list? Normally it shouldn't be possible for the wrong types to slip through, but it may be possible depending on where type checks are being compiled.
```suggestion
                allOf(
                  isA<List<Map<String, String>>>(),
                  everyElement(isA<Map<String, String>>()),
                ));
```

File: lib/src/util/prop_conversion.dart (hunk 1 above)
Comment (greglittlefield-wf) [reply lib/src/util/prop_conversion.dart]: Oh I was thinking that the generics on `.map` would get inferred and passed to `unjsifyMapProp`, but it looks like they don't because of the `toList` call, so `dynamic` is getting used instead.

So I think the current implementation:
```dart
return value?.map((e) => unjsifyMapProp(e as JsMap))?.toList()?.cast();
```
is equivalent to:
```dart
return value?.map((e) => unjsifyMapProp<dynamic, dynamic>(e as JsMap))?.toList()?.cast<Map<K, V>>();
```

And what we really want is to pass the generics to `unjsifyMapProp` so the returned maps have the right type, which should also eliminate the need for `.cast()`:
```dart
return value?.map((e) => unjsifyMapProp<K, V>(e as JsMap))?.toList();
```

Comment (sydneyjodon-wk) [reply lib/src/util/prop_conversion.dart]: Oh makes sense! Updated that!

File: test/over_react/util/prop_conversion_test.dart [hunk 5]
```diff
@@ -512,6 +608,232 @@ main() {
...
+                // For this test, it's important to mutate the the value returned from the `builder.listOfProps` getter,
+                // and not the original Map passed into `builder.listOfProps`.
+                builder.listOfProps[0]['data-foo'] = 'bar';
+                expect(builder.listOfProps, [
+                  {'data-foo': 'bar'}
+                ]);
+              });
+
+              group('and then reading a nested value that gets unconverted:', () {
+                group('A nested map prop', () {
```
Comment (greglittlefield-wf) [inline test/over_react/util/prop_conversion_test.dart:688]: #nit
```suggestion
                group('A nested map list prop', () {
```

**Review by greglittlefield-wf — APPROVED** [review]
+10 pending passing CI

**Comment by sydneyjodon-wk** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
